x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/action/ModelImporter.java

```diff
@@ -35,61 +36,173 @@
  */
 class ModelImporter {
     private static final int DEFAULT_CHUNK_SIZE = 1024 * 1024; // 1MB
+    private static final int MAX_IN_FLIGHT_REQUESTS = 3;
     private static final Logger logger = LogManager.getLogger(ModelImporter.class);
     private final Client client;
     private final String modelId;
     private final ModelPackageConfig config;
     private final ModelDownloadTask task;
+    private final ExecutorService executorService;
 
-    ModelImporter(Client client, String modelId, ModelPackageConfig packageConfig, ModelDownloadTask task) {
+    ModelImporter(Client client, String modelId, ModelPackageConfig packageConfig, ModelDownloadTask task, ThreadPool threadPool) {
         this.client = client;
         this.modelId = Objects.requireNonNull(modelId);
         this.config = Objects.requireNonNull(packageConfig);
         this.task = Objects.requireNonNull(task);
+        this.executorService = threadPool.executor(MachineLearningPackageLoader.MODEL_DOWNLOAD_THREADPOOL_NAME);
     }
 
-    public void doImport() throws URISyntaxException, IOException, ElasticsearchStatusException {
-        long size = config.getSize();
+    public void doImport(ActionListener<AcknowledgedResponse> finalListener) {
 
-        // Uploading other artefacts of the model first, that way the model is last and a simple search can be used to check if the
-        // download is complete
-        if (Strings.isNullOrEmpty(config.getVocabularyFile()) == false) {
-            uploadVocabulary();
+        long size = config.getSize();
+        // simple round up
+        int totalParts = (int) ((size + DEFAULT_CHUNK_SIZE - 1) / DEFAULT_CHUNK_SIZE);
+        InputStream modelInputStream;
+        ModelLoaderUtils.VocabularyParts vocabularyParts = null;
 
-            logger.debug(() -> format("[%s] imported model vocabulary [%s]", modelId, config.getVocabularyFile()));
+        try {
+            URI uri = ModelLoaderUtils.resolvePackageLocation(
+                config.getModelRepository(),
+                config.getPackagedModelId() + ModelLoaderUtils.MODEL_FILE_EXTENSION
+            );
+            modelInputStream = ModelLoaderUtils.getInputStreamFromModelRepository(uri);
```

Comment:
Looking at the upload server, it accepts range queries with the form `Range: bytes=0-1024` so we should be able to download concurrently (or read files concurrently using a random reader)? If downloading async is too complicated we can use the new download threadpool. That would also limit the number of download we have. Regarding the current blocking part, it happens on the ml utility thread pool, that's what worries me the most. I think we need to move the blocking part to the download thread pool

Reply:
With the range download we can have multiple input streams doing the download in parallel, each one doing a single range. Currently we limit the download to one input stream because we retrieve the entire file from one connection. With the range header we can use one connection for each part (limited to N connections in parallel with a dedicated thread pool if we want). 
I tried with the following:
```
curl --location 'https://ml-models.elastic.co/multilingual-e5-small.pt' \
--header 'Range: bytes=0-1024'
```
That returns the first 1024 bytes so we can split into multiple tasks, each responsible to download and store a range of the file instead of splitting from a single input stream?